### PR TITLE
PP-162: scheduler keeps waiting even after session id has arrived fro…

### PR DESCRIPTION
**Pull request by nithinj**
…m job launch

#### Issue-ID

* **[PP-162](https://pbspro.atlassian.net/browse/PP-162)**

#### Problem description
* Scheduler keeps waiting even after session id has arrived from job launch

#### Cause / Analysis
* wait_request() does not abort when a signal occurred, delaying send_job
* Sched sends job to server, waiting to be acked (before continuing on next job)
* send_job process sends job to mom. Mom acks server, send_job child process exits. Mom then starts process, sends session id to server. Server writes the start record.
* But it does not yet register the send_job child process exit, since its busy in the wait_request() loop serving IFL clients, and so does not reap the child process.
*  This keeps scheduler waiting for an extra second, resulting in (even) slower scheduling rates.

#### Solution description
* The implementation involves catching signals and handling them as soon as possible. There are two sections where the changes are done.
1. poll for data on socket
Convert poll to ppoll on platforms that support ppoll. ppoll helps to atomically unblock signals and look for data in the sockets. poll has a chance of race condition, using ppoll removes that chance. However ppoll is not supported on older versions of Linux and hence those flavors continue to use poll. A new define, HAVE_PPOLL is introduced in configure configure to make this differentiation.
2. while loop that handles socket data
The while loop handling socket data could iterate based on the number of sockets having data. The number of iterations could be more if there are many active clients. So we allow a small window where, if a signal is received, we ubblock for a small window and allow the signal to be handled, and exit from the while loop.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

CC: @subhasisb , @crjayadev

### CODE REVIEW

File: m4/pbs_decl_epoll_pwait.m4 [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#
+
+AC_DEFUN([PBS_AC_DECL_EPOLL_PWAIT],
+[
+  AS_CASE([x$target_os],
+    [xlinux*],
+      AC_MSG_CHECKING(whether epoll_pwait system call is supported)
+      AC_TRY_RUN(
+[
+#define _GNU_SOURCE
```
Comment (mike0042) [inline m4/pbs_decl_epoll_pwait.m4]: We don't define _GNU_SOURCE anywhere in PBS. Doing so in the check for epoll_pwait() is probably not the right thing to do.

File: m4/pbs_decl_ppoll.m4 [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#
+
+AC_DEFUN([PBS_AC_DECL_PPOLL],
+[
+  AS_CASE([x$target_os],
+    [xlinux*],
+     AC_MSG_CHECKING(whether ppoll API is supported)
+      AC_TRY_RUN(
+[
+#define _GNU_SOURCE
```
Comment (mike0042) [inline m4/pbs_decl_ppoll.m4]: We don't define _GNU_SOURCE anywhere in PBS. Doing so in the check for ppoll() is probably not the right thing to do.

File: src/include/job.h [hunk 3]
```diff
@@ -371,19 +371,20 @@ typedef	struct	hnodent {
...
 	long long	vn_mem;		/* working set size (real mem)	*/
 	long long	vn_vmem;	/* total mem space (virtual)	*/
-	int		vn_naccels;	/* number of accelerators */
-	int		vn_need_accel;	/* should we reserve the accelerators */
-	char	       *vn_accel_model;	/* model of the desired accelerator */
+	int 		vn_naccels;	/* number of accelerators */
+	int 		vn_need_accel;	/* should we reserve the accelerators */
+	char 		*vn_accel_model;	/* model of the desired accelerator */
 	long long	vn_accel_mem;	/* amt of accelerator memory wanted */
 } vmpiprocs;
```
Comment (subhasisb) [inline src/include/job.h]: None of the changes to job.h is related to this bug. We can discard these changes - probably showing due to the way you did the diff for the older changelist

File: src/include/net_connect.h [hunk 4]
```diff
@@ -208,5 +215,7 @@ struct connection {
 	void 		*cn_data;         /* pointer to some data for cn_func */
 	char		cn_username[PBS_MAXUSER];
 	char		cn_hostname[PBS_MAXHOSTNAME+1];
+	pbs_list_link cn_link;  /* link to the next connection in the linked list */
```
Comment (subhasisb) [inline src/include/net_connect.h]: please fix indentation

File: src/lib/Libtpp/tpp_em.c [hunk 5]
```diff
@@ -691,12 +753,28 @@ tpp_em_del_fd(void *em_ctx, int fd)
...
+		if (sigprocmask(SIG_SETMASK, sigmask, &origmask) == -1)
+			return -1;
+	}
+
+	if (sigmask) {
+		sigprocmask(SIG_SETMASK, &origmask, NULL);
+	}
+#endif
 
 	nready = poll(ctx->fds, ctx->curr_nfds, timeout);
```
Comment (subhasisb) [inline src/lib/Libtpp/tpp_em.c]: The poll() call should be in between the blocking and unblocking the signals using sigprocmask

File: src/lib/Libtpp/tpp_em.c [hunk 6]
```diff
@@ -972,10 +1104,11 @@ tpp_em_wait(void *em_ctx, em_event_t **ev_array, int timeout)
 		}
 
 		if (ev_count > ctx->max_nfds)
-			break;
+		break;
```
Comment (subhasisb) [inline src/lib/Libtpp/tpp_em.c]: indentation

File: src/resmom/start_exec.c [hunk 7]
```diff
@@ -3992,6 +3988,7 @@ job_nodes_inner(struct job *pjob, hnodent **mynp)
 	int	 need_accel = 0;  /* accelerator needed in subchunk? */
 	long long accel_mem = 0;  /* accel mem per exec_vnode key-value pair */
 	char 	*accel_model = NULL; /* accelerator model if set */
+	long	nppus = 0;
```
Comment (subhasisb) [inline src/resmom/start_exec.c]: This nppus change does not seem related to this bug fix

**Comment by nithinj** [comment]
Hi @mike0042 and @subhasisb ,
I've addressed your review comments. Please check.

**Comment by nithinj** [comment]
Hi @crjayadev ,
Ran automated tests in both the modes and could not found any new failure.

**Comment by nithinj** [comment]
Hi @mike0042 / @subhasisb ,
I've squashed and re-based my commit. Please take this further.

**Comment by subhasisb** [comment]
@nithinj looks like code errors introduced by this change. Please see https://pbspro.atlassian.net/browse/PP-1187
